src/lib/encryption.ts

```diff
@@ -0,0 +1,77 @@
+import { Poseidon, Group, Field, Scalar } from '../snarky';
+import { PrivateKey, PublicKey } from './signature';
+
+export { encrypt, decrypt };
+
+class Sponge {
+  sponge: unknown;
+
+  constructor() {
+    this.sponge = (Poseidon as any).spongeCreate()();
+  }
+
+  absorb(x: Field) {
+    // console.log(this.sponge, x);
+    (Poseidon as any).spongeAbsorb(this.sponge, x);
+  }
+
+  squeeze() {
+    return (Poseidon as any).spongeSqueeze(this.sponge);
+  }
+}
+
+Poseidon.Sponge = Sponge;
+
+type CipherText = {
+  publicKey: Group;
+  cipherText: Field[];
+};
+
+function encrypt(message: Field[], otherPublicKey: PublicKey) {
+  // key exchange
+  let privateKey = Scalar.random();
+  let publicKey = Group.generator.scale(privateKey);
+  let sharedSecret = otherPublicKey.g.scale(privateKey);
+
+  let sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x); // don't think we need y, that's enough entropy
+
+  // encryption
+  let cipherText = [];
+  for (let messageChunk of message) {
+    let keyStream = sponge.squeeze();
+    let encryptedChunk = messageChunk.add(keyStream);
+    sponge.absorb(encryptedChunk); // absorb for the auth tag
+    cipherText.push(encryptedChunk);
+  }
```

Comment:
just a thought, but this might be inefficient as theoretically we can absorb two field elements without permuting (the rate of our sponge is 2), but this squeeze-then-absorb might trigger a permutation for every field element. I'm not sure what's the implementation anymore.

Reply:
The relevant implementation is here: https://github.com/o1-labs/snarky/blob/master/sponge/sponge.ml#L287

As you say, the present version triggers one permutation per iteration - because we permute when absorbing and then squeezing. Without the in-between absorbing we'd only permute every second iteration.

Reply:
we might want to create an issue to optimize this at some point :o (or you might want to process chunks two field elements by two field elements)